### PR TITLE
12572 SP/GP Correction Edit UI - Update Business Name

### PR DESCRIPTION
**Pull request by lewischenstudio**
*Issue #:* [/bcgov/entity#12572](https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/issues/bcgov/entity/12572)

*Description of changes:*
-  reuse the existing component
- only support "new NR" version for name change
- added `isEntityTypeFirm` in name-request-mixin.ts

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Comment by severinbeauvais** [comment]
> only support "new NR" version for name change

Did you confirm that this is the actual requirement? I think staff should be able to correct company names (same as with BEN IA corrections). In any case, you didn't change this in this PR (it's configurable in the resource files), so please update either way.

**Comment by severinbeauvais** [comment]
You should update the app version.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:24]: This isn't the NR type. It's the type of company name correction being done.

Ref: https://github.com/bcgov/business-edit-ui/blob/main/src/enums/correctionTypes.ts

I think this method/mixin shouldn't care about the type of correction being done. It should only focus on NR. Therefore, I think it would be better if you passed in the information you really care about: is this is correction with a new NR?

Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts:24]: Actually, is this method called in any scenario other than for a name change NR? It's not clear to me.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:45]: OK, I'm pretty confused here. What is this variable exactly? Is it ever false? How does it affect the logic below -- isn't it always a new NR that we're validating?

I'm also confused about the scenario when either of the 2 following blocks is called.

Please add comments or review the variable names or review the scenarios.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:93]: I don't think we use a "new" NR for changing a business name. I believe we use a "change" NR, like this:

image

Please confirm with PO.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:318]: These unit test changes now only validate a GP/NEW NR and no longer validate a BEN CNV NR. Please add a unit test (or parameterize it) instead.

**Comment by lewischenstudio** [comment]
> > only support "new NR" version for name change
> 
> Did you confirm that this is the actual requirement? I think staff should be able to correct company names (same as with BEN IA corrections). In any case, you didn't change this in this PR (it's configurable in the resource files), so please update either way.

1. This is the [Design UI](https://preview.uxpin.com/363a49b33bdb2a314ccac0f18daad8969e93b337#/pages/150032525/simulate/sitemap). Dev is now unable to use the NR number, so the entire ticket is solve this problem. 
2. The method [`validateNameRequest`](https://github.com/bcgov/business-edit-ui/blob/a0d42e7fe7f782e2c5a3d6932a563df2117ff4ed/src/mixins/name-request-mixin.ts#L20-L65) validates the NR with the backend. Before this PR, it doesn't check for `CorrectionTypes.CORRECT_NEW_NR` while [`onSubmit`](https://github.com/bcgov/business-edit-ui/blob/a0d42e7fe7f782e2c5a3d6932a563df2117ff4ed/src/components/common/YourCompany/CompanyName/CorrectNameRequest.vue#L147-L158) uses `validateNameRequest` only when `this.formType === CorrectionTypes.CORRECT_NEW_NR`.
3. The way I go with this is to preserve the method `validateNameRequest` in its original way by adding an additional, conditional property `nrType`, which takes the value from `CorrectionTypes`. If this method is used other place than `onSubmit`, then it is still good.

**Comment by severinbeauvais** [comment]
> > Did you confirm that this is the actual requirement? I think staff should be able to correct company names (same as with BEN IA corrections). In any case, you didn't change this in this PR (it's configurable in the resource files), so please update either way.
> 
> 1. This is the [Design UI](https://preview.uxpin.com/363a49b33bdb2a314ccac0f18daad8969e93b337#/pages/150032525/simulate/sitemap). Dev is now unable to use the NR number, so the entire ticket is solve this problem.

Yes, and I expressed concern that the design was incorrect, and I asked you to confirm the requirement with the BAs. I still think staff should be able to correct the business name as with BEN IA corrections.

This PR is blocked until this requirement is confirmed.

**Comment by severinbeauvais** [comment]
> 2\. The method [`validateNameRequest`](https://github.com/bcgov/business-edit-ui/blob/a0d42e7fe7f782e2c5a3d6932a563df2117ff4ed/src/mixins/name-request-mixin.ts#L20-L65) validates the NR with the backend. Before this PR, it doesn't check for `CorrectionTypes.CORRECT_NEW_NR` while [`onSubmit`](https://github.com/bcgov/business-edit-ui/blob/a0d42e7fe7f782e2c5a3d6932a563df2117ff4ed/src/components/common/YourCompany/CompanyName/CorrectNameRequest.vue#L147-L158) uses `validateNameRequest` only when `this.formType === CorrectionTypes.CORRECT_NEW_NR`.

I don't understand this response.

**Comment by severinbeauvais** [comment]
> 4\. This ticket is for SP/GP corrections. I think BEN IA corrections have their tickets.

You have changed a method that is also used by BEN IA. You have also removed the BEN IA unit test. You do need to consider the other things you might be impacting in this UI.

**Comment by severinbeauvais** [comment]
> 3\. The way I go with this is to preserve the method `validateNameRequest` in its original way by adding an additional, conditional property `nrType`, which takes the value from `CorrectionTypes`. If this method is used other place than `onSubmit`, then it is still good.

Please see if you can refactor a little bit to clarify this code. It's very confusin and unclear what scenario it handles. (Comments would help.) And again, the mixin should not know what happens outside its scope, such as what kind of form it's called from. Instead, if you can't refactor this out, rename the parameter to tell the method that it should do X instead of Y. I'm hoping you can refactor and greatly simplify the logic here and make it easily comprehensible. Thanks.

**Review by severinbeauvais — CHANGES_REQUESTED** [review]
See comments above, especially:
- requirement for different name change types
- what kind of NR is needed (NEWvs CNV)
- confusing logic
- previous unit test essentially removed

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:66]: See other comment re: NEW not a valid NR for this; should be CHANGE OF NAME.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts]: This PR is blocked until this requirement is confirmed.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:67]: Could this code change accidentally allow a "new" NR for a BEN IA correction, which previously didn't allow this type?

File: src/components/common/YourCompany/YourCompany.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/YourCompany.vue:469]: You should use the resource files to do this, not special-case code.

We want the components to be as generic as possible. We have resource files to specify a lot of things.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts]: Latest code has special case for firms so it no longer impacts others.

**Comment by severinbeauvais** [comment]
> > > Did you confirm that this is the actual requirement? I think staff should be able to correct company names (same as with BEN IA corrections). In any case, you didn't change this in this PR (it's configurable in the resource files), so please update either way.
> > 
> > 
> > 
> > 1. This is the [Design UI](https://preview.uxpin.com/363a49b33bdb2a314ccac0f18daad8969e93b337#/pages/150032525/simulate/sitemap). Dev is now unable to use the NR number, so the entire ticket is solve this problem.
> 
> Yes, and I expressed concern that the design was incorrect, and I asked you to confirm the requirement with the PO. I still think staff should be able to correct the business name as with BEN IA corrections.
> 
> This PR is blocked until this requirement is confirmed.

Confirmed by Linda that we are presenting just 1 option for firm corrections: use a new name request.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts]: Confirmed by Linda that a "new" name request cannot be used in a firm correction (and not a "conversion" NR either). Only a "change of name" NR can be used.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts]: Confirmed by Linda that a "new" name request cannot be used in a firm correction (and not a "conversion" NR either). Only a "change of name" NR can be used.

File: src/components/common/YourCompany/YourCompany.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/YourCompany.vue:469]: No. We're trying to keep components generic (ie, to know as little as possible about the different variations depending on entity type).

As I have mentioned before, there are resource files that specify the name change options for each entity type.

File: src/components/common/YourCompany/YourCompany.vue
Comment (lewischenstudio) [reply src/components/common/YourCompany/YourCompany.vue]: This is now changed.

File: src/mixins/name-request-mixin.ts
Comment (lewischenstudio) [reply src/mixins/name-request-mixin.ts]: It is now using "CHG" for firm filings.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:63]: So now this mixin needs to know about entity type and has a special case for firms.

It would be better if the acceptable NR types were stored in the resource files (and if they were passed in to this method).

File: tests/unit/name-request-mixin.spec.ts
Comment (severinbeauvais) [inline tests/unit/name-request-mixin.spec.ts:175]: Wrong request type code for a firm NR.

Comment (severinbeauvais) [reply tests/unit/name-request-mixin.spec.ts:175]: Sorry, I see you're testing the "false" case here.

However, I don't see where you set a firm request type code in here, so you're not actually testing firm NR validation.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:400]: I'm thinking lines 398 and 400 should be SP.

I'm also wondering whether the mixin should verify that the type code matches the current filing type...? Or can a GP NR be used for a SP correction? Please check with PO/BA and report back. Thanks.

File: src/components/common/YourCompany/YourCompany.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/YourCompany.vue:472]: Are the resource objects already available in the store? Search for setResource and getResource.

Comment (severinbeauvais) [reply src/components/common/YourCompany/YourCompany.vue:472]: Ditto in mixin below.

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [inline src/mixins/name-request-mixin.ts:64]: It would be good to use the resource files for this case also (which I think is only Benefit Companies but please verify).

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts]: This change now looks good. 👍

File: src/components/common/YourCompany/YourCompany.vue
Comment (severinbeauvais) [reply src/components/common/YourCompany/YourCompany.vue]: Looks good.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts]: ^^ outstanding question to answer

File: src/components/common/YourCompany/YourCompany.vue
Comment (severinbeauvais) [inline src/components/common/YourCompany/YourCompany.vue:469]: Lines 468 and 469 return the same value, so you don't need the first one.

Comment (lewischenstudio) [reply src/components/common/YourCompany/YourCompany.vue:469]: ok

Comment (severinbeauvais) [reply src/components/common/YourCompany/YourCompany.vue:468]: Did you have something else in mind? If so, please add a code comment explaining it.

Comment (lewischenstudio) [reply src/components/common/YourCompany/YourCompany.vue:468]: No, I didn't. This makes sense now.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (lewischenstudio) [reply tests/unit/CorrectNameRequest.spec.ts]: GP NR cannot be used for a SP correction. They must be the same type: GP NR for GP correction and SP NR for SP correction.

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts]: OK. Can you implement a type check in the NR mixin, or create a new ticket?

File: src/mixins/name-request-mixin.ts
Comment (severinbeauvais) [reply src/mixins/name-request-mixin.ts:62]: Faiz's PR was ready and merged first, but his work is not complete and he's aware of your changes here, so he'll update this for coop special resolutions in his next PR.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (lewischenstudio) [reply tests/unit/CorrectNameRequest.spec.ts]: I'll implement the type check in this PR.

Comment (lewischenstudio) [reply tests/unit/CorrectNameRequest.spec.ts]: After I implemented the type check, then I found out that there is already an entity type checker implemented: [here](https://github.com/bcgov/business-edit-ui/blob/5457c482ab6d6fbee20060f14525f29824f7b224/src/components/common/YourCompany/CompanyName/CorrectNameRequest.vue#L159-L176). If you think this is fine, I will revert my code and then push it.

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts]: Oh, oops.

Yes, I think the original type check is fine.

**Review by severinbeauvais — APPROVED** [review]
GTG after you fix the merge conflicts / update the app version.

@lambert-alex Can you review this too, please?

**Comment by lewischenstudio** [comment]
> GTG after you fix the merge conflicts / update the app version.
> 
> @lambert-alex Can you review this too, please?

Thanks for the reviews 🙏

**Comment by severinbeauvais** [comment]
Ref: /bcgov/entity#12572

**Pull request merged**
